C++: Port dataflow tests to inline expectations test library.

criemen opened the pull request:
Fixes https://github.com/github/codeql-c-analysis-team/issues/158 by converting our dataflow tests to `InlineExpectationTests`.

MathiasVP commented on cpp/ql/test/library-tests/dataflow/dataflow-tests/test.ql:
```diff
@@ -1,5 +1,27 @@
-from DataFlow::Node sink, DataFlow::Node source, TestAllocationConfig cfg
-where cfg.hasFlow(source, sink)
-select sink, source
+class ASTDataFlowTest extends InlineExpectationsTest {
```
This class (and the IR version) looks a lot like the `ASTFieldFlowTest` class we have in the `fields` directory. At some point, it'd be nice to unify them (but that doesn't have to be in this PR)

MathiasVP commented on cpp/ql/test/library-tests/dataflow/dataflow-tests/dispatch.cpp:
```diff
@@ -20,29 +20,29 @@ struct Middle : Top {
-  topPtr->isSink(source()); // flow [NOT DETECTED by AST]
+  sink(topPtr->isSource1()); // $ ir MISSING: ast
+  sink(topPtr->isSource2()); // $ ir MISSING: ast
+  topPtr->isSink(source()); // causing a MISSING for ast
```
Is there a reason this one is not an inline expectation comment (and similarly for the others below this one)?

criemen replied:
Yes indeed, as the sink function is not called here, but in the implementation of `Bottom`, this is not the correct target for the inline expectation comment. However I still wanted to annotate the caller here with something. Let me know if/how this could be made more obvious.

MathiasVP replied:
Ah, now I understand. Thanks!

MathiasVP reviewed (commented):
Looks great!

jbj commented on cpp/ql/test/library-tests/dataflow/dataflow-tests/clang.cpp:
```diff
@@ -15,39 +15,38 @@ void following_pointers(
   sink(sourceArray1[0]); // no flow
   sink(*sourceArray1); // no flow
-  sink(&sourceArray1); // flow (should probably be taint only)
+  sink(&sourceArray1); // $ ast // [should probably taint only]
```
```suggestion
  sink(&sourceArray1); // $ ast // [should probably be taint only]
```

jbj reviewed (commented):
LGTM so far!

criemen commented:
The new commits contain:
- porting all of taint- tests
- deletion of the test_diff tests (the information is now inline in all the test files and can be found by grep)
- vscode snippets for annotating the tests

The last two points might be something folks want to argue about.

criemen commented:
Note: I was surprised by the large amount of false negatives with the IR taint tracking when it comes to STL classes - our models work for AST taint tracking, but the IR lags behind.
For investigation, you can have a look at (for example) `taint-tests/vector.cpp` and all the lines with the ` ast MISSING: ir` annotation.
The other STL models perform similarly.
Maybe this is already a known short coming of IR-based taint/data flow, then just ignore this remark.

There is also this IR FP that stands out, because it is really difficult here for us to explain to the user why there should be data flow:
https://github.com/github/codeql/pull/4741/files#diff-f2917de01b533d5fde19348a47b7f23ce8cd72b9cee0babc890856910a44ce7eR92

MathiasVP commented:
> Note: I was surprised by the large amount of false negatives with the IR taint tracking when it comes to STL classes - our models work for AST taint tracking, but the IR lags behind.

We still need some model parts of flow (see for instance the TODO here: https://github.com/github/codeql/blob/main/cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll#L714). It was improved a lot by https://github.com/github/codeql/pull/4295, but there's still a way to go.

MathiasVP commented on cpp/ql/test/library-tests/dataflow/taint-tests/map.cpp:
```diff
@@ -151,104 +151,104 @@ void test_map()
-		sink(i2->first); // tainted
-		sink(i2->second); // [FALSE POSITIVE]
+		sink(*i3); // $ MISSING: ast,ir
+		sink(i3->first);  // $ MISSING: ast,ir
+		sink(i3->second); // clean
```
Well spotted!

MathiasVP commented on .vscode/InlineExpectationsTests.code-snippets:
```diff
@@ -0,0 +1,60 @@
+{
```
We haven't added any `.code-snippets` in the codeql repo yet, and I'm not aware of a policy regarding this. While I'm personally a big fan of this particular snippet, I don't know if the inline expectation test framework has enough of a userbase to justify putting it in the repo. Do you have any opinion about this?

criemen replied:
I expected this discussion to happen - my main counterpoint is that these snippets are restricted to cpp files in our repo, which is only the tests, so most of our users will never see the snippets.
Those who do may or might not work on dataflow/taint tracking tests and then they may or may not find them useful.

I mainly wanted to share these snippets internally, as I believe they will be useful for everybody on the team, given that we evolve dataflow/taint tracking regularly.

jbj replied:
If one person other than @criemen finds the snippets useful, then I won't object to checking them in. I'm not that person: I think the burden of typing these annotations out will be far smaller when most of them have been converted and we only need to add annotations for new cases or sometimes edit existing ones.

MathiasVP reviewed (commented):
LGTM! I only have a single comment regarding the `.code-snippets`.

jbj commented:
Should this PR be Ready for review now?

geoffw0 commented on cpp/ql/test/library-tests/dataflow/taint-tests/map.cpp:
```diff
@@ -303,136 +303,136 @@ void test_unordered_map()
-		sink(i2->second); // [FALSE POSITIVE]
+		sink(*i3); // $ MISSING: ast,ir
+		sink(i2->first); // $ ir MISSING: ast
+		sink(i2->second); // $ SPURIOUS: ir
```
`i2` should be `i3` here as in the similar case further up.

geoffw0 reviewed (commented):
This looks great. Now I'm kind of pleased we had several rounds of bikeshedding what the inline expectations should look like. :)

Also thank you for your patience with the huge number of test cases in these files.

MathiasVP commented on cpp/ql/test/library-tests/dataflow/DefaultTaintTracking/annotate_path_to_sink/dispatch.cpp:
```diff
@@ -0,0 +1,74 @@
+void notSink(int notSinkParam);
+
+void callsSink(int sinkParam) {
+  sink(sinkParam); // $ ast,ir=31:28 ast,ir=32:31 ast,ir=34:22 MISSING: ast,ir=28
```
That `MISSING` seems to be missing a line/column number, right?

criemen commented:
The PR is now ready to review, as all 3 directories mentioned in https://github.com/github/codeql-c-analysis-team/issues/158 are converted.

To summarize, I want to draw the reviewers attention to the following details:
- For DefaultTaintTracking, I split tests into two groups - one where annotate all nodes on the path to a sink, and one where we annotate sink invocations only. This is different what we did previous, where we dumped out all elements that were reachable from a taint source - regardless of whether that was on a path to a sink or not. This change is needed, as otherwise the test files are unreadable. For example in `sink(a.str())` we'd have to annotate this with `ast,ir` if `a` is tainted, even if `a.str()` isn't.
- As mentioned above, this commit contains the (debatable) addition of VSCode snippets
- I deleted all difference tests between ast and ir, as I feel that the state of that is sufficiently represented in the test files. If an overview is needed, a sufficiently advanced grep can produce that.
- I feel that the `DefaultTaintTracking` tests, and especially the one for global variables are particular weird, but then I am not always sure about the expected test outcome.
- There will be a follow-up PR cleaning up the test QL code

MathiasVP commented on cpp/ql/test/library-tests/dataflow/DefaultTaintTracking/annotate_path_to_sink/tainted.ql:
```diff
@@ -0,0 +1,88 @@
+/**
+ * This test provides the possibility to annotate elements when they are on a path of a taint flow to a sink.
```
I think this is a good idea! I guess this works because `DefaultTaintTracking` has a very generous definition of sinks. It might be good to have a separate tag for nodes that are part of the path, but is not the actual sink?

criemen replied:
I don't understand your comment.
Indeed, this feature defines and filters on a custom notion of sink (namely `sink` invocations).
Can you try to explain differently what you mean?
I don't feel for the current tests that we need different tags for `on-the-path-but-not-the-sink` and `the-sink`.

However, an annotation `tainted-but-not-on-a-path-to-a-sink` is doable - I just doubt it's usefulness, but am open to be convinced otherwise.

MathiasVP replied:
Yeah on second thought I don't think we gain anything by using separate tags. This LGTM!

MathiasVP commented on cpp/ql/test/library-tests/dataflow/DefaultTaintTracking/annotate_sinks_only/defaulttainttracking.cpp:
```diff
@@ -156,35 +117,35 @@ void sink(Point);
-  sink(p); // tainted [field -> object]
-  sink(&p); // tainted [field -> object]
-  sink(p.x); // tainted
+  sink(p); // $ MISSING: ast,ir // tainted [field -> object]
```
Oh yeah. The lines with `tainted [field -> object]` comments are actually supposed to missing. They are examples of flow from field to object (which we intend to remove as part of https://github.com/github/codeql-c-analysis-team/issues/103), and https://github.com/github/codeql/pull/4230 removed these results (but apparently I didn't update the annotations, sorry!).

criemen replied:
Ah, good to know, I'll remove all the MISSING annotations then.

MathiasVP reviewed (commented):
LGTM apart from a few comments!

MathiasVP reviewed (approved):
LGTM!
